rust/rsc/src/common/database.rs

```diff
@@ -157,7 +190,226 @@ pub async fn read_dbonly_blob_store(
+    db: &T,
+    output_symlinks: Vec<output_symlink::ActiveModel>,
+) -> Result<(), DbErr> {
+    if output_symlinks.len() == 0 {
+        return Ok(());
+    }
+
+    let chunked: Vec<Vec<output_symlink::ActiveModel>> = output_symlinks
+        .into_iter()
+        .chunks(MAX_SQLX_PARAMS / 5)
+        .into_iter()
+        .map(|chunk| chunk.collect())
+        .collect();
+
+    for chunk in chunked {
+        OutputSymlink::insert_many(chunk).exec(db).await?;
+    }
+
+    Ok(())
+}
+
+// ----------             Read             ----------
+
+// ----------            Update            ----------
+
+// ----------            Delete            ----------
+
+// --------------------------------------------------
+// ----------          Output Dir          ----------
+// --------------------------------------------------
+
+// ----------            Create            ----------
+pub async fn create_many_output_dirs<T: ConnectionTrait>(
+    db: &T,
+    output_dirs: Vec<output_dir::ActiveModel>,
+) -> Result<(), DbErr> {
+    if output_dirs.len() == 0 {
+        return Ok(());
+    }
+
+    let chunked: Vec<Vec<output_dir::ActiveModel>> = output_dirs
+        .into_iter()
+        .chunks(MAX_SQLX_PARAMS / 5)
+        .into_iter()
+        .map(|chunk| chunk.collect())
+        .collect();
+
+    for chunk in chunked {
+        OutputDir::insert_many(chunk).exec(db).await?;
+    }
+
+    Ok(())
+}
+
+// ----------             Read             ----------
+
+// ----------            Update            ----------
+
+// ----------            Delete            ----------
+
+// --------------------------------------------------
+// ----------             Blob             ----------
+// --------------------------------------------------
+
+// ----------            Create            ----------
+
+// ----------             Read             ----------
+
+// Reads blobs from the database that are unreferenced and have surpassed the allocated grace
+// period to be referenced.
+//
+// For new blobs this allows the client to create several blobs and then reference them all at
+// once. Existing blobs whose job was just evicted will likely be well past the grace period and
+// thus quickly evicted themselves.
+pub async fn read_unreferenced_blobs<T: ConnectionTrait>(
+    db: &T,
+    ttl: NaiveDateTime,
+) -> Result<Vec<blob::Model>, DbErr> {
+    // TODO: this can probably be refactored as a ORM query with something like
+    // Blob::find()
+    //     .filter(
+    //         blob::Column::Id.not_in_subquery(
+    //             sea_orm::query::Query::select()
+    //                 .column(OutputFile::BlobId)
+    //                 .from(OutputFile::Table
+    //                 .union( *OTHER STUFF* ));
+    //                 .take().
+
+    // Limit = 16k as the query is also subject to parameter max.
+    // Blob has 4 params so (2^16)/4 = 16384. Also generally best to chunk blob eviction
+    // to avoid large eviction stalls.
```

Comment:
How frequently does the eviction loop run? Given that this function only returns "some" of the unreferenced blobs if we are trying to evict more than 16k in between invocations we will run into problems. If eviction is running frequently enough I think there will be other more pertinent limiters but just want to check.

Reply:
Its something we can configure but right now its set to tick every minute. So as long as we don't add more than 16k blobs per minute averaged across the whole day then it should keep up